### PR TITLE
fix: keeps hooks intact when attaching backend

### PR DESCRIPTION
**Pull request by akshbhu**
#### Description of changes
- Since from v10.0.0 hooks aren't maintained by amplify. During amplify pill operation when attaching backend , hooks pulled from git are getting deleted by amplify when doing a pull in an empty folder or running headless pull command
```amplify pull --envName dev```
Fix here is to keeps the hooks back into amplify folder when attaching backend.

#### Description of how you validated changes
- tested manually
- Added an e2e test for the specific case 
- cloud -e2e : https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli/11792/workflows/f5312d44-f86a-4dfb-944e-a9ea523263d0
#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/amplify-cli/src/attach-backend.ts [hunk 1]
```diff
@@ -100,7 +100,10 @@ const onSuccess = async (context: $TSContext): Promise<void> => {
   } else if (stateManager.currentMetaFileExists()) {
     await initializeEnv(context, stateManager.getCurrentMeta());
   }
-
+  // move Hooks folder from backup to original amplify folder
+  const HooksDirPath = pathManager.getHooksDirPath(projectPath);
+  const HooksBackupDirPath = path.join(backupAmplifyDirPath, 'hooks');
```
Comment (sobolk) [inline packages/amplify-cli/src/attach-backend.ts:105]: nit: lowercase ?

File: packages/amplify-cli/src/attach-backend.ts [hunk 2]
```diff
@@ -100,7 +100,10 @@ const onSuccess = async (context: $TSContext): Promise<void> => {
...
+  fs.moveSync(HooksBackupDirPath, HooksDirPath, { overwrite: true });
```
Comment (sobolk) [inline packages/amplify-cli/src/attach-backend.ts:106]: 1. is it possible that `hooks` doesn't exist at all?
2. Should we aways overwrite if it exists at the destination ?

Comment (akshbhu) [reply packages/amplify-cli/src/attach-backend.ts:106]: 1. Nice catch! , I can include a check but usually we  add a hooks folder during init .
2. Yes , There us no hooks present at destination as we already put that in backup amplify holder while generating backend.

File: packages/amplify-e2e-core/src/init/amplifyPull.ts [hunk 3]
```diff
@@ -116,7 +116,7 @@ export const amplifyPullNonInteractive = (
       configLevel: 'project',
       useProfile: true,
       // eslint-disable-next-line spellcheck/spell-checker
-      profileName: 'amplify-integ-test-user',
+      profileName: 'default',
```
Comment (akshbhu) [inline packages/amplify-e2e-core/src/init/amplifyPull.ts:119]: change this also

**Review by sobolk — APPROVED** [review]

**Review by lazpavel — APPROVED** [review]

File: yarn.lock [hunk 4]
```diff
@@ -9195,7 +9195,7 @@
 
 "@types/http-cache-semantics@*":
   version "4.0.1"
-  resolved "https://registry.npmjs.org/@types/http-cache-semantics/-/http-cache-semantics-4.0.1.tgz#0ea7b61496902b95890dc4c3a116b60cb8dae812"
+  resolved "https://registry.yarnpkg.com/@types/http-cache-semantics/-/http-cache-semantics-4.0.1.tgz#0ea7b61496902b95890dc4c3a116b60cb8dae812"
```
Comment (edwardfoyle) [inline yarn.lock:9198]: I think the changes to this file should be reverted

**Comment by hackmajoris** [comment]
Waiting for this PR to be merged. I hope it will fix this: https://github.com/aws-amplify/amplify-cli/issues/11021

File: scripts/split-e2e-tests.ts [hunk 5]
```diff
@@ -35,7 +35,7 @@ const WINDOWS_TEST_ALLOWLIST: string[] = [
   'schema-auth-11-c_pkg',
   'auth_6_pkg',
   'frontend_config_drift_pkg',
-  'hooks_pkg',
+  'hooks-b_pkg',
```
Comment (evcodes) [inline scripts/split-e2e-tests.ts:38]: where is hooks-a?

Comment (awsluja) [reply scripts/split-e2e-tests.ts:38]: good catch, should include both a/b here

Comment (akshbhu) [reply scripts/split-e2e-tests.ts:38]: This windows allow list , hooks_a file wont run on windows, only on linux

File: packages/amplify-e2e-tests/src/__tests__/hooks-b.test.ts [hunk 6]
```diff
@@ -1,26 +1,10 @@
...
+  addFunction, createNewProjectDir,
   deleteProject,
-  deleteProjectDir,
-  getAppId,
-  getBucketKeys,
-  getHooksDirPath,
-  getProjectMeta,
-  initJSProjectWithProfile,
-  removeFunction,
+  deleteProjectDir, getHooksDirPath, initJSProjectWithProfile,
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/hooks-b.test.ts:4]: nit: can we format these into new lines?

**Review by evcodes — CHANGES_REQUESTED** [review]
I have few nits but otherwise looks good to merge

File: scripts/split-e2e-tests.ts (hunk 5 above)
Comment (evcodes) [reply scripts/split-e2e-tests.ts:38]: ok then I'll resolve

**Review by evcodes — APPROVED** [review]

**Review by sobolk — DISMISSED** [review]

File: packages/amplify-e2e-tests/src/__tests__/hooks-a.test.ts [hunk 7]
```diff
@@ -0,0 +1,37 @@
+import {
+  addFunction, amplifyPullNonInteractive, amplifyPushAuth, createNewProjectDir,
```
Comment (evcodes) [inline packages/amplify-e2e-tests/src/__tests__/hooks-a.test.ts:2]: these too 😭

**Review by sobolk — APPROVED** [review]

**Review by evcodes — APPROVED** [review]

**Comment by hackmajoris** [comment]
@akshbhu any idea when this fixed will be released?

**Pull request merged**